Add Diagnosticable base class and add documentation.

jacob314 opened the pull request.

jacob314 committed:
Add Diagnosticable base class and documentation.

jacob314 commented on packages/flutter/lib/src/material/mergeable_material.dart:
```diff
@@ -122,10 +122,10 @@ class MergeableMaterial extends StatefulWidget {
   final bool hasDividers;
 
   @override
-  String toString() {
```
I assume this toString was added by accident and an int value is displayed with toStringAsFixed by accident as well.
The base class should be adding the key if it exists so there is no need to add it again here.
 Correct me if I'm wrong and I'll look at adding an option to display ints using toStringAsFixed. E.g. so elevation shows as 3.0 instead of 3.

Hixie replied:
elevation should be a `double`, if it's an int that's an error

jacob314 replied:
in this case it does indeed need to be an int for the internals of the class to work correctly. Switched the diagnostic view of the property to show as a DoubleProperty as the fact that the value can only be an int shouldn't be visible in the debugging output.

Hixie replied:
k. Please file a bug about this, elevation should be a double everywhere.

jacob314 commented on packages/flutter/lib/src/painting/box_painter.dart:
```diff
@@ -1601,24 +1601,29 @@ class BoxDecoration extends Decoration {
 
+  @protected
   @override
-  DiagnosticsNode toDiagnosticsNode({ String name, DiagnosticsTreeStyle style: DiagnosticsTreeStyle.whitespace }) {
```
The meat of this CL is these refactors from using DiagnosticsNode.lazy to just extending Diagnosticable or DiagnosticableTree.

jacob314 commented on packages/flutter/lib/src/painting/text_span.dart:
```diff
@@ -248,11 +248,6 @@ class TextSpan implements TreeDiagnostics {
     return result;
   }
 
-  @override
```
Switched to the standard behavior of a TreeDiagnostics class where toString generates a full single line description and fixed the only usage that called toString passing in an argument.

jacob314 commented on packages/flutter/lib/src/rendering/flex.dart:
```diff
@@ -776,8 +776,8 @@ class RenderFlex extends RenderBox with ContainerRenderObjectMixin<RenderBox, Fl
   @override
-  String toString() {
-    String header = super.toString();
+  String toShortDescription() {
```
I switched all the RenderObject classes to have a toString that returns a full description of the object and properties. If this is a mistake I can switch it back but having the toString methods for RenderObject be inconsistent with other Diagnosticable objects seems odd.

Hixie replied:
Did you change all the places that call toString on RenderObjects as well? It's a very common thing to implicitly do in error messages (using string interpolation).

Hixie replied:
I think we're probably better off having the toString return the short description of a RenderObject, because RenderObjects' string descriptions don't generally fit on one line but the render object toString is frequently used in the middle of paragraphs and such.

jacob314 replied:
I've switched the toString for RenderObject back. However, it does seem like a lot of error messages were already using toStringShort or toStringDeep rather than toString for RenderObject.
However, i don't know a reliable way in IntelliJ to find all usages for RenderObject.toString so I can't say for sure.

Hixie replied:
Many of the places that use `$this` in rendering/*.dart, for one, but also any time anyone wants to do printf debugging for a render object and just sticks "`$this`" in their print statement.

jacob314 replied:
for printf debugging I would think seeing the properties would usually be extremely helpful but maybe I'm wrong about that. This CL is already big enough so perhaps in a followup I'll look at all the $this in rendering/.*dart etc.

Hixie replied:
For widgets it's often helpful. For render objects it's usually more helpful to know the identity of the render object than its precise state at the time of printing, though obviously both are sometimes useful (hence toStringShallow).

jacob314 commented on packages/flutter/lib/src/rendering/proxy_box.dart:
```diff
@@ -1485,7 +1485,7 @@ class RenderDecoratedBox extends RenderProxyBox {
             'Every call to save() or saveLayer() must be matched by a call to restore().\n'
             'The decoration was:\n'
-            '${decoration.toStringDeep()}\n'
+            '  $decoration\n'
```
the regular toString for a decoration is now just fine. Having the method be toStringDeep for a decoration seems odd as a decoration doesn't have children.

jacob314 commented on packages/flutter/test/foundation/diagnostics_test.dart:
```diff
@@ -274,29 +274,16 @@ void main() {
-      '  multi\n'
-      '  line\n'
-      '  value!\n',
+        'TestTree#00000(stringProperty1: value1; doubleProperty1: 42.5; roundedProperty: 0.3; nullProperty: null; <root node>)\n'
```
I had misunderstood how the dense style should look and the dense widget tree was working previously by adding all the properties on the side to the description instead of through the standard property syntax.

jacob314 commented on packages/flutter/test/widgets/framework_test.dart:
```diff
@@ -486,24 +486,23 @@ void main() {
- └LimitedBox(maxWidth: 0.0; maxHeight: 0.0; renderObject: RenderLimitedBox#00000 relayoutBoundary=up1)
-  └ConstrainedBox(BoxConstraints(biggest); renderObject: RenderConstrainedBox#00000 relayoutBoundary=up2)
-''',
+        'Column([GlobalKey#00000]; renderObject: RenderFlex#00000)\n'
```
Change here is Container
is displayed instead of
Container()
I can make the Dense style put parens around empty parameter lists but that seems odd.
I am assuming placing parens around empty param lists was just an oversight as the similar existing singleLine style did not place parens around empty parameter lists. Just let me know. It is trivial to switch this default back.

Hixie replied:
I could go either way on this. Let's try to live with your change and see if the inconsistency is made up for by the cleaner output.

jacob314 replied:
sgtm

jacob314 commented on packages/flutter/test/rendering/box_test.dart:
```diff
@@ -76,7 +76,7 @@ void main() {
         '   constraints: null\n'
         '   size: MISSING\n'
-        '   decoration:\n'
+        '   decoration: BoxDecoration:\n'
```
Previously the name of the decoration class used was omitted. I can make BoxDecoration class use '' as its description 
but that seems a bit more like a bug than a feature.
When you call toString on a box decoration I would think you definitely want 
BoxDecoration(<no decorations specified>)
instead of
<no decorations specified> but perhaps the idea is to try to keep toString for decorations very terse in which case I can switch back to the original behavior.

Hixie replied:
yeah, this seems like an improvement

jacob314 replied:
good

jacob314 commented on packages/flutter/test/foundation/diagnostics_test.dart:
```diff
@@ -534,11 +519,8 @@ void main() {
-        ' │ ║   └child node B3: TestTree#00000\n'
-        ' │ ║     <leaf node>\n'
-        ' │ ║     foo: 42\n'
+        ' │ ║   ├child dense: TestTree#00000(property1: "value1")\n'
```
just fixing cases where I hadn't internalized what the dense style meant.

Hixie commented on packages/flutter/lib/src/foundation/diagnostics.dart:
```diff
@@ -158,10 +160,13 @@ class TextTreeConfiguration {
   /// [singleLineTextConfiguration].
   final String lineBreak;
 
+  /// Whether to line break properties or to leave all properties on one line.
```
clarify the docs by answering: Does this mean "put line breaks between properties" or "wrap properties with line breaks"?

jacob314 replied:
Done. Place line breaks between properties.

Hixie commented on packages/flutter/lib/src/foundation/diagnostics.dart:
```diff
@@ -2108,10 +2028,72 @@ abstract class TreeDiagnosticsMixin implements TreeDiagnostics {
+  }
+
+  @override
+  String toShortDescription() => describeIdentity(this);
```
why not toStringShort?

jacob314 replied:
Because toStringShort was too good a name so there are existing conflicting and not really compatible toStringShort methods defined by
Recognizer 
e.g. 
String toStringShort() => 'horizontal multidrag';

And RenderObjectToWidgetAdapter
  String toStringShort() => debugShortDescription ?? super.toStringShort();

I'm open to moving those toStringShort methods over to a different name if you would prefer that.
I would think the RenderObjectToWidgetAdapter should probably be eliminated or refactored anyway. I have less context on the Recognizer case.

Hixie replied:
I think it's fine for those two classes to override toStringShort in that way, colliding with them is fine. We should just view toStringShort as a way to represent the object tersely. Some classes will want to provide more useful versions of that than just their identity.

jacob314 replied:
Done. Please look at the test output diff to make sure the output is reasonable. I kept this change as its own commit on my branch so it is easy to view in isolation.
I'm now hiding the key property from Widget as it is redundant. New output looks generally reasonable.

Hixie commented on packages/flutter/lib/src/foundation/diagnostics.dart:
```diff
@@ -2108,10 +2028,72 @@ abstract class TreeDiagnosticsMixin implements TreeDiagnostics {
+/// A class that can be used as a mixin that helps dump string and
+/// [DiagnosticsNode] representations of trees.
+///
+/// Use this class as a mixin in any context where it is possible to use mixins.
```
these docs say they're for a mixin but you have a very similar mixin class so presumably these are to be a base class?

jacob314 replied:
Fix mixin mixup. Cleaned up the doc for this Diagnosticable and DiagnosticableTree.

Hixie commented on packages/flutter/lib/src/foundation/diagnostics.dart:
```diff
@@ -1796,114 +1762,68 @@ String camelCaseToHyphenatedName(String word) {
-    return toDiagnosticsNode().toStringDeep(prefixLineOne, prefixOtherLines);
+  /// Default diagnostics tree style to use.
+  @protected
+  DiagnosticsTreeStyle get debugDefaultDiagnosticsTreeStyle {
```
since this class' members will appear on a whole bunch of APIs, I would rather we kept the number of members (even protected ones) to a strict minimum. Ideally just toString, toStringShort (named that so that it sorts after and next to toString), toDiagnosticsNode (which is inevitable since it's the point of the class), and debugFillProperties (which is the one thing we want people to override).

jacob314 replied:
done. Used your idea from debugFillProperties

Hixie commented on packages/flutter/lib/src/painting/box_painter.dart:
```diff
@@ -1601,24 +1601,29 @@ class BoxDecoration extends Decoration {
+  String toShortDescription() => '$runtimeType';
+
+  @override
+  void debugFillProperties(List<DiagnosticsNode> properties) {
```
one option for getting rid of debugDefaultDiagnosticsTreeStyle and debugEmptyBodyDescription is for `properties` here to not be a `List<DiagnosticsNode>` but instead be an object on which you can call `add`, and then to which we add the feature that you can set its `emptyBodyDescription` and `defaultTreeStyle` or whatnot.

jacob314 replied:
Brilliant! That is much better than the alternative of forcing subclasses to override toDiagnosticNode.

Hixie commented on packages/flutter/lib/src/foundation/diagnostics.dart:
```diff
@@ -2126,7 +2108,66 @@ abstract class TreeDiagnosticsMixin implements TreeDiagnostics {
+  @override String get debugEmptyBodyDescription => null;
+
+  @override
+  @protected
```
having both override and protected seems like an error. Only the place that introduces the member should say protected. If you have override, you shouldn't have anything else.

jacob314 replied:
done

Hixie commented on packages/flutter/lib/src/foundation/diagnostics.dart:
```diff
@@ -2126,7 +2108,66 @@ abstract class TreeDiagnosticsMixin implements TreeDiagnostics {
+
+  @override
+  @protected
+  @mustCallSuper
```
same here. either override, or protected/mustCallSuper.

jacob314 replied:
done

Hixie commented on packages/flutter/lib/src/foundation/diagnostics.dart:
```diff
@@ -2108,10 +2028,72 @@ abstract class TreeDiagnosticsMixin implements TreeDiagnostics {
+    return new _DiagnosticableTreeNode(
+      name: name,
+      value: this,
+      style: style ?? debugDefaultDiagnosticsTreeStyle,
```
never mind, see other comments

Hixie commented:
This is overall a great improvement, just some small things to tweak.

jacob314 commented:
Pushed new commits to the branch to review. Addressed all comments. Ready for the next round of review.

Hixie commented on packages/flutter/lib/src/cupertino/page.dart:
```diff
@@ -424,19 +424,12 @@ class _CupertinoEdgeShadowDecoration extends Decoration {
-        properties.add(new DiagnosticsProperty<LinearGradient>('edgeGradient', edgeGradient));
-      },
-    );
+  String toStringShort() => '$runtimeType';
```
isn't this handled by the Decoration base class now?

jacob314 replied:
I don't currently have the decoration base class override toStringShort. I could do that if it is believed that no decoration subclasses should use the default className#hashCode toStringShort

One theory is it is more idiomatic to specify that the short string is $runtimeType only on the class that implements operator ==, etc.

Hixie replied:
all the decoration subclasses should act the same, so I think it's safe to make Decoration do whatever the BoxDecoration / FlutterLogoDecoration / _CupertinoEdgeShadowDecoration classes do (hopefully they're all consistent).

Hixie commented on packages/flutter/lib/src/foundation/diagnostics.dart:
```diff
@@ -584,44 +579,17 @@ abstract class DiagnosticsNode {
-    this.emptyBodyDescription,
-  }) : assert(style != null),
-       assert(showName != null),
+  }) : assert(showName != null),
```
If style is allowed to be null, update the docs...

Hixie commented on packages/flutter/lib/src/foundation/diagnostics.dart:
```diff
@@ -2108,10 +2020,82 @@ abstract class TreeDiagnosticsMixin implements TreeDiagnostics {
+  void debugFillProperties(DiagnosticPropertiesBuilder properties) { }
+}
+
+/// An base class for providing string and [DiagnosticsNode] debug
```
A

jacob314 replied:
done

Hixie commented on packages/flutter/lib/src/painting/text_span.dart:
```diff
@@ -276,7 +271,7 @@ class TextSpan implements TreeDiagnostics {
           'A TextSpan object with a non-null child list should not have any nulls in its child list.\n'
           'The full text in question was:\n'
-          '${toString("  ")}'
+          '${toStringDeep()}'
```
won't this no longer be indented now?

jacob314 replied:
Done. Indented all lines by 2 spaces.

Hixie commented on packages/flutter/test/gestures/debug_test.dart:
```diff
@@ -25,7 +25,7 @@ void main() {
     expect(log, hasLength(2));
     expect(log[0], equalsIgnoringHashCodes('Gesture arena 1    ❙ ★ Opening new gesture arena.'));
-    expect(log[1], equalsIgnoringHashCodes('Gesture arena 1    ❙ Adding: TapGestureRecognizer#00000(state: ready)'));
+    expect(log[1], equalsIgnoringHashCodes('Gesture arena 1    ❙ Adding: tap(state: ready)'));
```
argh, yeah, that's no good. We need the hashcode here so we can recognize it in the logs.

Ok, how about for the recognizers we have an entirely separate "debugDescription" getter that's totally independent of the diagnostics stuff, which we use for the short name, and anywhere where previously we used toStringShort, we now use that?

Hixie commented on packages/flutter/test/widgets/table_test.dart:
```diff
@@ -526,7 +526,7 @@ void main() {
       element.toStringDeep(),
       equalsIgnoringHashCodes(
-        'Table([GlobalKey#00000]; renderObject: RenderTable#00000)\n'
+        'Table-[GlobalKey#00000](renderObject: RenderTable#00000)\n'
```
this is going to take some getting used to but it looks ok

Hixie commented:
LGTM modulo the Decoration issue and the Recognizer issue.

jacob314 commented:
Addressed Decoration and Recognizer issues.

Hixie commented:
![LGTM](http://junkyard.damowmow.com/577)

jacob314 committed:
Fixed example code that was broken in the process of refactoring.

jacob314 committed:
Pass the configuration for the parent down the tree enabling
better formatting of properties when the parent displays the properties
as a single line.
Updated test expectations to match the change.

Hixie commented on packages/flutter/test/rendering/debug_test.dart:
```diff
@@ -30,14 +31,8 @@ void main() {
-            '  [3] 0.0,0.0,0.0,1.0',
-      ),
+      simple.toString(parentConfiguration: singleLineTextConfiguration),
+      equals('transform: [2.0,0.0,0.0,0.0; 0.0,2.0,0.0,0.0; 0.0,0.0,2.0,0.0; 0.0,0.0,0.0,1.0]'),
```
can we verify that in anything that isn't explicitly single-line, this still looks like a matrix?

Hixie commented:
LGTM

jacob314 committed:
Fix code review comment

jacob314 committed:
Update one new debugFillProperties method to use the new signature.

Hixie commented on packages/flutter/lib/src/rendering/flex.dart:
```diff
@@ -489,11 +489,8 @@ class RenderFlex extends RenderBox with ContainerRenderObjectMixin<RenderBox, Fl
-              for (DiagnosticsNode line in description)
-                information.writeln('  $line');
+              information.write('  ');
+              information.write(node.toStringShallow('\n  '));
```
Turns out this hides a regression: we're now missing a newline at the end of the ancestor print out, and the "See also" on the next line ends up on the last line:

```
...
I/flutter ( 4533): The nearest ancestor providing an unbounded width constraint is:
I/flutter ( 4533):   _RenderSingleChildViewport#76a9a relayoutBoundary=up10 NEEDS-LAYOUT NEEDS-PAINT
I/flutter ( 4533):   creator: _SingleChildViewport ← _ScrollableScope ← IgnorePointer-[GlobalKey#4a64a] ← Semantics ←
I/flutter ( 4533):   Listener ← _GestureSemantics ←
I/flutter ( 4533):   RawGestureDetector-[LabeledGlobalKey<RawGestureDetectorState>#20712] ←
I/flutter ( 4533):   _ExcludableScrollSemantics-[GlobalKey#fc34f] ← RepaintBoundary ← CustomPaint ← RepaintBoundary ←
I/flutter ( 4533):   NotificationListener<ScrollNotification> ← ⋯
I/flutter ( 4533):   parentData: <none> (can use size)
I/flutter ( 4533):   constraints: BoxConstraints(0.0<=w<=484.0, 0.0<=h<=920.0)
I/flutter ( 4533):   layer: OffsetLayer#77c9c
I/flutter ( 4533):   size: Size(484.0, 920.0)See also: https://flutter.io/layout/
I/flutter ( 4533): If this message did not help you determine the problem, consider using debugDumpRenderTree():
I/flutter ( 4533):   https://flutter.io/debugging/#rendering-layer
I/flutter ( 4533):   http://docs.flutter.io/flutter/rendering/debugDumpRenderTree.html
...
```

Hixie replied:
I'll submit a PR to fix it.
